### PR TITLE
driver: eth: gmac: scale down the TX path

### PR DESCRIPTION
**Pull request by aurel32**
This PR downscale the TX path following the discussion in PR #12563. It means that the GMAC driver is now safe with regard to this PR, even in the PTP case.

### CODE REVIEW

**Review by pfalcon — COMMENTED** [review]
@aurel32: I appreciate your looking into these matters again. But as I explained in https://github.com/zephyrproject-rtos/zephyr/pull/12563, I in no way call for massive code removals. By "scale down" I mean "add temporal short-circuit code paths to "skip" some too-advanced features of the driver which don't play well with the current state of the IP stack".

Unfortunately, what I see here more looks like "massive code removals" rather than "lightweight, easily revertible changes".

**Comment by aurel32** [comment]
> @aurel32: I appreciate your looking into these matters again. But as I explained in #12563, I in no way call for massive code removals. By "scale down" I mean "add temporal short-circuit code paths to "skip" some too-advanced features of the driver which don't play well with the current state of the IP stack".
> 
> Unfortunately, what I see here more looks like "massive code removals" rather than "lightweight, easily revertible changes".

I can try to just to wrap the code that is removed with `#if 0 #endif`. Now I don't think only adding a mutex or semaphore is a good idea given it means a lot of flash space, RAM space and CPU cycles will just be wasted for no gain. In addition the goal of scaling down the driver is to be able is to keep the drivers simple so that we can fix the stack. If that is really the goal, I am not sure a driver referencing and unreferencing packet and fragments from the ISR is a good idea, even under mutex.

To finish, if we agree that the current IP stack doesn't support DMA, it means things are allowed to change quite a lot, and I am not sure than once it supports DMA again, the code will still be usable. Can you predict how access to the content of the packet from multiple threads will be handled? At the packet level? At the fragment level? Other?

**Comment by aurel32** [comment]
An additional note: if we need to restore the code later, we can always use the git history for that.

**Comment by pfalcon** [comment]
> Can you predict how access to the content of the packet from multiple threads will be handled?

As a quick reply, access to the same resource from multiple threads shouldn't just happen. That's the golden rule. An alternative is to have clear ownership rules - and that's exactly the background of what we were talking https://github.com/zephyrproject-rtos/zephyr/pull/12563: when stack calls driver's ->send(), the driver owns the packet. When ->send() returns, the driver no longer owns packet. That's not flexible enough, but that's very clear rule, and effectively what we have now. Let's make that processing work without a hitch, then see how to extend it.

The worst alternative is when we add piles of locks everywhere to make those multiple threads bite at the same resource. It doesn't happen too much though, as soon we start to get deadlocks.

**Comment by pfalcon** [comment]
> An additional note: if we need to restore the code later, we can always use the git history for that.

I never by that argument, sorry ;-).

However, I buy following argument: If we write code destined for removal, then instead we can sit and wait - we'll arrive to the same point ;-). So, assuming we keep that in mind, and write code for other purpose than removal, let's value the code (especially written by others) and maintain it (which may involve temporary short-circuiting/disabling it). And we can also write less, but better code ;-).

**Comment by aurel32** [comment]
> > Can you predict how access to the content of the packet from multiple threads will be handled?
> 
> As a quick reply, access to the same resource from multiple threads shouldn't just happen. That's the golden rule. An alternative is to have clear ownership rules - and that's exactly the background of what we were talking #12563: when stack calls driver's ->send(), the driver owns the packet. When ->send() returns, the driver no longer owns packet. That's not flexible enough, but that's very clear rule, and effectively what we have now. Let's make that processing work without a hitch, then see how to extend it.

So if we do not know more about how it will be, it means we have no idea if the existing code will be usable or not. It won't be just flipping a bit to just re-enable it.

**Comment by pfalcon** [comment]
> So if we do not know more about how it will be, it means we have no idea if the existing code will be usable or not.

I'm sure we will be able. Heck, that's our job as programmers. ;-) And it will be *much* easier than digging some old code in git history.

Anyway, that's all generic comments, to find a common ground in approach. Being more specific would require more detailed review. Which in turn being backlogged by RC hitting on us and more mundane (depends on PoV apparently) bugs hitting on us. So, what the real current state of this. Are you doing this on the hot steps of discussion in #12563, or is it still required to stabilize gmac driver for 1.14 (I got impression that various other related things were committed, which might alleviate the situation) ?

**Comment by aurel32** [comment]
@pfalcon I have just pushed a new version that keeps the old version of the code. Switching between one version and the other is just a matter of flipping a define.

**Comment by mnkp** [comment]
Considering that lack of support for DMA Ethernet drivers is temporary I strongly agree with @pfalcon and his proposal
> The main point - no need to remove any code from GMAC driver. Just need to introduce a blocking on mutex/sema at the end of its send() - i.e. return not when DMA request is set up and started, but when it's finished. The DMA itself will be used, and later the whole matter can be revisited.

In fact I would rather go the other direction, make synchronous / asynchronous mode of gmac driver configurable via a (possibly hidden) Kconfig option or define. The thing is, it's very difficult to come up with a new improved architecture when there is no way to verify the design. Having gmac driver work synchronously by default but being able to easily switch to asynchronous mode to test new implementation is going to be invaluable. Unless networking stack is going to drop support for DMA Ethernet drivers completely. I hope that's not going to happen.

**Comment by mnkp** [comment]
I see i was too late writing my comment. @aurel32 thanks for the update!

**Comment by aurel32** [comment]
> In fact I would rather go the other direction, make synchronous / asynchronous mode of gmac driver configurable via a (possibly hidden) Kconfig option or define.

I have done that in the latest version. It can be switched with a define.

> Unless networking stack is going to drop support for DMA Ethernet drivers completely. I hope that's not going to happen.

To be clear DMA is still supported in the RX path, and DMA for a single packet is still supported in the TX path. However the current networking stack doesn't support queueing multiple TX packets anymore. It works at low traffic rate, but the `net_buf` reference accounting is broken with high traffic rate as it is not thread safe, so the whole stack ends up locking up at some point. In addition PTP also doesn't work in that case.

**Review by tgorochowik — APPROVED** [review]
+1 from me, just because this PR + #13204 render the gptp sample on sam e-70 functional again.

**Review by jukkar — APPROVED** [review]

File: drivers/ethernet/eth_sam_gmac.c [hunk 1]
```diff
@@ -227,24 +225,23 @@ static void mac_addr_set(Gmac *gmac, u8_t index,
...
+	__ASSERT_NO_MSG(rx_frag_list);
 
 	rx_desc_list->tail = 0U;
-	rx_frag_list->tail = 0U;
 
 	for (int i = 0; i < rx_desc_list->len; i++) {
 		rx_buf = net_pkt_get_reserve_rx_data(K_NO_WAIT);
 		if (rx_buf == NULL) {
-			free_rx_bufs(rx_frag_list);
+			free_rx_bufs(rx_frag_list, rx_desc_list->len);
```
Comment (mnkp) [inline drivers/ethernet/eth_sam_gmac.c]: I think we can pass `i` as length. Then there will also be no need to have `rx_frag_list[i] = NULL;` in `free_rx_bufs`.

**Review by mnkp — APPROVED** [review]

**Pull request merged**